euphonic/design/06_brille_integration.md

```diff
@@ -0,0 +1,199 @@
+adding symmetry information to Euphonic's `Crystal` object (so that Brille
+doesn't need to find it itself).
+
+**Example Syntax**
+
+From Python:
+
+```
+from euphonic import ureg, ForceConstants
+from euphonic.util import mp_grid
+
+fc = ForceConstants.from_castep('quartz.castep_bin')
+
+# Create BrilleForceConstants object
+interpolate_kwargs = {'use_c': True, 'n_threads': 4, 'asr': 'reciprocal'}
+bfc = fc.calculate_brille_force_constants(
+    grid='mesh', max_points=10000, **interpolate_kwargs)
+
+# Create DebyeWaller
+dw_phonons = bfc.calculate_qpoint_phonon_modes(mp_grid([6,6,6]))
+dw = dw_phonons.calculate_debye_waller(temperature=300*ureg('K'))
+
+# Calculate StructureFactor
+phonons = bfc.calculate_qpoint_phonon_modes(qpts, n_threads=4)
+sf = phonons.calculate_structure_factor(dw)
+```
+
+From Matlab:
+
+The question here is whether the interface should be 'Horace-like' with bespoke
+functions to make it easier to simulate cuts, or whether it should more closely
+mirror Euphonic's Python interface. The following example is somewhere
+inbetween and assumes a Matlab wrapper has been written (named 'meuphonic').
+
+```
+% Create ForceConstants and BrilleForceConstants using Euphonic-like syntax
+fc = meuphonic.ForceConstants.from_castep('quartz.castep_bin');
+kwargs = {'use_c', true, 'n_threads', 4};
+bfc = fc.calculate_brille_force_constants( ...
+    'grid', 'mesh', 'max_points', 10000, 'interpolate_kwargs', kwargs);
+
+fit_params = [1.0]; % temperature, scale
+pars = {fit_params,
+        'force_constants', bfc,  % accepts fc or bfc
+        'interpolate_kwargs', {'threads', 4}, % different depending on fc or bfc
+        'debye_waller_grid', [6,6,6],
+        'temperature', 300.0,
+        'conversion_mat', [1,0,0; 0,1,0; 0,0,-1], % Convert Horace Q -> Euphonic Q
+        'chunk', 10000};
+sim_cut = disp2sqw_eval(exp_cut, @meuphonic.horace_sqw, pars, fwhh);
+```
+
+**Advantages:**
+ - Would benefit both Horace and Euphonic users
+ - Allows Horace users to skip a step depending on if they want to use brille
+   or not
+
+**Disadvantages:**
+ - Would require additions to Euphonic's ForceConstants, and quite a few
+   changes to Brille's existing repos before this option is feasible
+ - The interface could end up being quite different to SpinW (does this
+   matter?)
+
+### Use case 2: Brille Q-point Mesh for BZ Integration
+This would involve using Brille to produce a mesh of q-points and their weights
+which Euphonic could use for Brillouin Zone integration (instead of naively
+generating a Monkhorst-Pack grid using `mp_grid` as in previous examples, which
+makes no use of symmetry). This is a much 'lighter' use of Brille and should in
+theory be easy to implement. However, libraries that Euphonic already depends
+on such as `spglib` can already generate weighted q-point Monkhorst-Pack grids,
+so is there actually much to be gained from having a 'smarter' mesh generated
```

Comment:
```
so is there actually much to be gained from having a 'smarter' mesh generated
```
@krefson @ajjackson please correct me if I'm wrong, but one of the 'bad' things about MP grids is that they are always constructed from linear combinations of the basis vectors. So if the angle between two basis vectors is unusually small (or large) or the ratio between any two basis vector lengths is large a MP grid may be poorly conditioned for some calculations.
The 'grids' constructed by `brille` are always built in an orthonormal coordinate system as either a Delaunay tetrahedralisation or from a series of (truncated) cubic cells.  So the positions of the points in a `brille` grid (away from the irreducible Brillouin zone boundaries) do not depend on the basis vectors.

Reply:
I think there's a fundamental question to be asked about the merits of linear interpolation for BZ integration. Is it actually going to improve the results?

The work on generalised grids from Gus Hart's group shows improvement over naive Monkhorst-Pack sampling for electronic structure of realistic materials. https://arxiv.org/pdf/1804.04741.pdf I'd expect this to transfer to phonon sampling as well. I'm not sure how Brille's method compares.

An earlier piece of important work in this area is Moreno/Soler https://journals.aps.org/prb/abstract/10.1103/PhysRevB.45.13891 ; they illustrate the impact of different primitive cell choices. They also recommend optimising the grid origin for maximum symmetry-reduction. This part hasn't been popular; it's no fun to implement and I for one prefer to keep the GP in my calculations because it's usually interesting.